source/common/websocket/codec.cc

```diff
@@ -0,0 +1,215 @@
+#include "source/common/websocket/codec.h"
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/scalar_to_byte_vector.h"
+
+namespace Envoy {
+namespace WebSocket {
+
+std::vector<uint8_t> Encoder::encodeFrameHeader(const Frame& frame) {
+  std::vector<uint8_t> output;
+  // Set flags and opcode
+  pushScalarToByteVector(
+      static_cast<uint8_t>(frame.final_fragment_ ? (0x80 | frame.opcode_) : frame.opcode_), output);
+
+  // Set payload length
+  if (frame.payload_length_ <= 125) {
+    // Set mask bit and 7-bit length
+    pushScalarToByteVector(frame.masking_key_ ? static_cast<uint8_t>(frame.payload_length_ | 0x80)
+                                              : static_cast<uint8_t>(frame.payload_length_),
+                           output);
+  } else if (frame.payload_length_ <= 65535) {
+    // Set mask bit and 16-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xfe : 0x7e), output);
+    // Set 16-bit length
+    pushScalarToByteVector(htobe16(frame.payload_length_), output);
+  } else {
+    // Set mask bit and 64-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xff : 0x7f), output);
+    // Set 64-bit length
+    pushScalarToByteVector(htobe64(frame.payload_length_), output);
+  }
+  // Set masking key
+  if (frame.masking_key_) {
+    pushScalarToByteVector(htobe32(frame.masking_key_.value()), output);
+  }
+  return output;
+}
+
+bool Decoder::doDecodeFlagsAndOpcode(uint8_t flags_and_opcode) {
+  // Validate opcode (last 4 bits)
+  uint8_t opcode = flags_and_opcode & 0x0f;
+  if (std::find(kFrameOpcodes.begin(), kFrameOpcodes.end(), opcode) == kFrameOpcodes.end()) {
+    return false;
+  }
+  frame_.opcode_ = opcode;
+  frame_.final_fragment_ = flags_and_opcode & 0x80;
+  state_ = State::FrameHeaderMaskFlagAndLength;
+  return true;
+}
+
+void Decoder::frameMaskFlag(uint8_t mask_and_length) {
+  // Set the mask length to be read
+  masking_key_length_ = mask_and_length & 0x80 ? kMaskingKeyLength : 0;
+  // Set length (0 to 125) or length flag (126 or 127)
+  length_ = mask_and_length & 0x7F;
+}
+
+void Decoder::frameDataStart(Buffer::Instance& input, absl::optional<std::vector<Frame>>& output) {
+  frame_.payload_length_ = length_;
+  if (length_ == 0) {
+    frameDataEnd(input, output);
```

Comment:
Instead of doing the "End" processing here, let's add an explicit Finished state. Then we can do:

```
void Decoder::frameDataStart() {
  frame_.payload_length_ = length_;
  if (length_ == 0) {
    state_ = State::FrameFinished;
  } else {
    frame_.payload_ = std::make_unique<Buffer::OwnedImpl>();
    state_ = State::FramePayload;
  }
}
```